faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -369,19 +374,66 @@ public static URI appendUri(URI oldUri, String queryKey, String queryValue) thro
     final Timer.Context ctx = registry.timer("fauna-request").time();
     final CompletableFuture<HttpResponse<String>> rv = new CompletableFuture<>();
 
+    HttpRequest request = makeHttpRequest(httpMethod, path, body, params, requestQueryTimeout, HttpClient.Version.HTTP_1_1);
+
+    client.sendRequest(request).whenCompleteAsync((response, throwable) -> {
+      ctx.stop();
+
+      if (throwable != null) {
+        logFailure(request, throwable);
+        rv.completeExceptionally(throwable);
+        return;
+      }
+
+      Optional<String> txnTimeHeader = response.headers().firstValue("X-Txn-Time");
+      txnTimeHeader.ifPresent(s -> syncLastTxnTime(Long.parseLong(s)));
+
+      logSuccess(request, response);
+
+      rv.complete(response);
+    });
+
+    return rv;
+  }
+
+  public CompletableFuture<HttpResponse<Flow.Publisher<List<ByteBuffer>>>> performStreamRequest(String httpMethod, String path, JsonNode body,
+                                                                                                final Optional<Duration> requestQueryTimeout) throws Exception {
+    final Timer.Context ctx = registry.timer("fauna-request").time();
+    final CompletableFuture<HttpResponse<Flow.Publisher<List<ByteBuffer>>>> rv = new CompletableFuture<>();
+
+    HttpRequest request = makeHttpRequest(httpMethod, path, Optional.of(body), Map.of(), requestQueryTimeout, HttpClient.Version.HTTP_2);
+
+    client.streamRequest(request).whenCompleteAsync((response, throwable) -> {
+      ctx.stop();
+      if (throwable != null) {
+        logFailure(request, throwable);
+        rv.completeExceptionally(throwable);
+        return;
+      }
+
+      Optional<String> txnTimeHeader = response.headers().firstValue("X-Txn-Time");
+      txnTimeHeader.ifPresent(s -> syncLastTxnTime(Long.parseLong(s)));
+
+      rv.complete(response);
+    });
+
+    return rv;
+  }
+
+  private HttpRequest makeHttpRequest(String httpMethod, String path, Optional<JsonNode> body, Map<String, String> params,
```

Comment:
`makeHttpRequest` was introduced as helper to decrease duplication of request building, it does not look great in the diff. though.